wgengine/magicsock/magicsock.go

```diff
@@ -1239,7 +1246,38 @@ func (c *Conn) GetEndpointChanges(peer tailcfg.NodeView) ([]EndpointChange, erro
 
 // DiscoPublicKey returns the discovery public key.
 func (c *Conn) DiscoPublicKey() key.DiscoPublic {
-	return c.discoPublic
+	return c.discoPublic.Load()
+}
+
+// RotateDiscoKey generates a new discovery key pair and updates the connection
+// to use it. This invalidates all existing disco sessions and will cause peers
+// to re-establish discovery sessions with the new key.
+//
+// This method is primarily intended for testing and debugging.
+//
+// Rotations are serialized to ensure only one rotation happens at a time.
+func (c *Conn) RotateDiscoKey() {
+	c.rotateDiscoMu.Lock()
+	defer c.rotateDiscoMu.Unlock()
+
+	oldShort := c.discoShort.Load()
+	newPrivate := key.NewDisco()
+	newPublic := newPrivate.Public()
+	newShort := newPublic.ShortString()
```

Comment:
Discussion (non-blocking): This seems wildly unlikely, but should we retry if the new short is the same as the old one? I would definitely do a double-take by a log line that didn’t appear to change things. Although maybe this shrinks the keyspace in a way that someone could exploit? I am OK with either outcome.